Can't read with "file" from inside User Op

Repro is with Zed commit 90db56f.

The following works ok without User Ops.

```
$ zq -version
Version: v1.8.1-27-g90db56f9

$ cat file_main.zed
file sample.zng | count()

$ zq -I file_main.zed
31(uint64)
```

However, if I try to move the logic to inside a User Op, now it won't get past the CLI tooling.

```
$ cat file_in_op.zed
op CallMe(): (
  file sample.zng | count()
)
CallMe()

$ zq -I file_in_op.zed
no input specified: use a command-line file or a Zed source operator
```

---

Incidentally, while working my way up to that repro, at one point I had this variation with a parameter that caused a panic.

```
$ cat file_in_op_with_param.zed
op CallMe(const param): (
  file sample.zng | count()
)
CallMe("goodbye")

$ echo "null" | zq -I file_in_op_with_param.zed -
panic: runtime error: slice bounds out of range [:-1]
goroutine 1 [running]:
runtime/debug.Stack()
	/usr/local/opt/go/libexec/src/runtime/debug/stack.go:24 +0x65
github.com/brimdata/zed/runtime/op.(*Catcher).Pull.func1()
	/Users/phil/work/zed/runtime/op/catcher.go:25 +0x45
panic({0x22a5ca0, 0xc0000ec420})
	/usr/local/opt/go/libexec/src/runtime/panic.go:884 +0x213
github.com/brimdata/zed/runtime/op.(*exitScope).Pull(0xc0000be078, 0x5?)
	/Users/phil/work/zed/runtime/op/scope.go:46 +0x10e
github.com/brimdata/zed/runtime/op.(*Single).Pull(0xc0000be408, 0x80)
	/Users/phil/work/zed/runtime/op/mux.go:134 +0x35
github.com/brimdata/zed/runtime/op.(*Catcher).Pull(0xc0004b8020?, 0x60?)
	/Users/phil/work/zed/runtime/op/catcher.go:28 +0x6d
github.com/brimdata/zed/runtime.(*Query).Pull(0x2b7a8360?, 0x30?)
	/Users/phil/work/zed/runtime/query.go:91 +0x44
github.com/brimdata/zed/zbuf.CopyPuller({0x2b7a8360, 0xc0008222e8}, {0x25c1460, 0xc000120cc0})
	/Users/phil/work/zed/zbuf/batch.go:165 +0x53
github.com/brimdata/zed/cli/zq.(*Command).Run(0xc000124380, {0xc00003e0b0, 0x1, 0x1})
	/Users/phil/work/zed/cli/zq/command.go:162 +0x798
github.com/brimdata/zed/pkg/charm.path.run({0xc0000a8278?, 0x1, 0x1}, {0xc00003e0b0?, 0x1, 0x0?})
	/Users/phil/work/zed/pkg/charm/path.go:11 +0x8b
github.com/brimdata/zed/pkg/charm.(*Spec).ExecRoot(0xc0000061a0?, {0xc00003e090, 0x3, 0x3})
	/Users/phil/work/zed/pkg/charm/charm.go:63 +0x4a
main.main()
	/Users/phil/work/zed/cmd/zq/main.go:11 +0x5d
```

It looks like I triggered what may be the same stack dump when attempting the [self `join` example from the tutorial](https://zed.brimdata.io/docs/next/tutorials/join#self-joins).

```
$ zq -version
Version: v1.8.1-28-g23140538

$ cat inner-join-streamed-in-op-with-params.zedop do_join(const left_match, const right_match, ...): (
  switch (
    case has(this[left_match]) => sort flavor
    case has(this[right_match]) => sort likes
  ) | inner join on flavor=likes eater:=name
)
do_join("color", "age", this)

$ cat fruit.ndjson people.ndjson | zq -z -I inner-join-streamed-in-op-with-params.zed -
panic: runtime error: slice bounds out of range [:-2]
goroutine 1 [running]:
runtime/debug.Stack()
	/usr/local/opt/go/libexec/src/runtime/debug/stack.go:24 +0x65
github.com/brimdata/zed/runtime/op.(*Catcher).Pull.func1()
	/Users/phil/work/zed/runtime/op/catcher.go:25 +0x45
panic({0x22a5be0, 0xc0000c21b0})
	/usr/local/opt/go/libexec/src/runtime/panic.go:884 +0x213
github.com/brimdata/zed/runtime/op.(*exitScope).Pull(0xc000932078, 0x5?)
	/Users/phil/work/zed/runtime/op/scope.go:46 +0x10e
github.com/brimdata/zed/runtime/op.(*Single).Pull(0xc000932090, 0xc0)
	/Users/phil/work/zed/runtime/op/mux.go:134 +0x35
github.com/brimdata/zed/runtime/op.(*Catcher).Pull(0xc0004ba020?, 0x70?)
	/Users/phil/work/zed/runtime/op/catcher.go:28 +0x6d
github.com/brimdata/zed/runtime.(*Query).Pull(0x2c9d4508?, 0x30?)
	/Users/phil/work/zed/runtime/query.go:91 +0x44
github.com/brimdata/zed/zbuf.CopyPuller({0x2c9d4508, 0xc000932438}, {0x25c1340, 0xc0009b8780})
	/Users/phil/work/zed/zbuf/batch.go:165 +0x53
github.com/brimdata/zed/cli/zq.(*Command).Run(0xc00002e380, {0xc0000ce040, 0x1, 0x1})
	/Users/phil/work/zed/cli/zq/command.go:162 +0x798
github.com/brimdata/zed/pkg/charm.path.run({0xc0000c4298?, 0x1, 0x1}, {0xc0000ce040?, 0x1, 0x0?})
	/Users/phil/work/zed/pkg/charm/path.go:11 +0x8b
github.com/brimdata/zed/pkg/charm.(*Spec).ExecRoot(0xc0000061a0?, {0xc0000ce010, 0x4, 0x4})
	/Users/phil/work/zed/pkg/charm/charm.go:63 +0x4a
main.main()
	/Users/phil/work/zed/cmd/zq/main.go:11 +0x5d
```

--- a/compiler/job.go
+++ b/compiler/job.go
@@ -40,17 +40,11 @@ func NewJob(octx *op.Context, in ast.Seq, src *data.Source, head *lakeparse.Comm
 	if len(seq) == 0 {
 		return nil, errors.New("internal error: AST seq cannot be empty")
 	}
-	from, reader, err := buildFrom(seq[0], head)
-	if err != nil {
-		return nil, err
-	}
-	if from != nil {
-		seq.Prepend(from)
-	}
 	entry, err := semantic.Analyze(octx.Context, seq, src, head)
 	if err != nil {
 		return nil, err
 	}
+	reader, _ := entry[0].(*kernel.Reader)
 	return &Job{
 		octx:      octx,
 		builder:   kernel.NewBuilder(octx, src),
@@ -58,71 +52,6 @@ func NewJob(octx *op.Context, in ast.Seq, src *data.Source, head *lakeparse.Comm
 		reader:    reader,
 		entry:     entry,
 	}, nil
-}
-
-func buildFrom(op ast.Op, head *lakeparse.Commitish) (*ast.From, *kernel.Reader, error) {
-	var from *ast.From
-	switch op := op.(type) {
-	case *ast.From:
-		// Already have an entry point with From.  Do nothing.
-		return nil, nil, nil
-	case *ast.Scope:
-		if len(op.Body) == 0 {
-			return nil, nil, errors.New("internal error: scope op has empty body")
-		}
-		return buildFrom(op.Body[0], head)
-	default:
-		var readers *kernel.Reader
-		trunk := ast.Trunk{Kind: "Trunk"}
-		if head != nil {
-			// For the lakes, if there is no from operator, then
-			// we default to scanning HEAD (without any of the
-			// from options).
-			trunk.Source = &ast.Pool{
-				Kind: "Pool",
-				Spec: ast.PoolSpec{
-					Pool: &ast.String{
-						Kind: "String",
-						Text: "HEAD",
-					},
-				},
-			}
-		} else {
-			readers = &kernel.Reader{}
-			trunk.Source = readers
-		}
-		from = &ast.From{
-			Kind:   "From",
-			Trunks: []ast.Trunk{trunk},
-		}
-		// XXX why not move this above? or
-		if isParallelWithLeadingFroms(op) {
-			from = nil
-			readers = nil
-		}
-		return from, readers, nil
-	}
-}
-
-func isParallelWithLeadingFroms(o ast.Op) bool {
-	par, ok := o.(*ast.Parallel)
-	if !ok {
-		return false
-	}
-	for _, seq := range par.Paths {
-		if !hasLeadingFrom(seq) {
-			return false
-		}
-	}
-	return true
-}
-
-func hasLeadingFrom(seq ast.Seq) bool {
-	if len(seq) == 0 {
-		return false
-	}
-	_, ok := seq[0].(*ast.From)
-	return ok
 }
 
 func (j *Job) Entry() dag.Seq {

--- a/compiler/kernel/op.go
+++ b/compiler/kernel/op.go
@@ -704,6 +704,8 @@ func isEntry(seq dag.Seq) bool {
 		return true
 	case *dag.Scope:
 		return isEntry(op.Body)
+	case *dag.UserOpCall:
+		return isEntry(op.Body)
 	case *dag.Fork:
 		if len(op.Paths) == 0 {
 			return false

--- a/compiler/semantic/analyzer.go
+++ b/compiler/semantic/analyzer.go
@@ -7,6 +7,7 @@ import (
 	"github.com/brimdata/zed/compiler/ast"
 	"github.com/brimdata/zed/compiler/ast/dag"
 	"github.com/brimdata/zed/compiler/data"
+	"github.com/brimdata/zed/compiler/kernel"
 	"github.com/brimdata/zed/lakeparse"
 )
 
@@ -14,7 +15,19 @@ import (
 // to DAG form, resolving syntax ambiguities, and performing constant propagation.
 // After semantic analysis, the DAG is ready for either optimization or compilation.
 func Analyze(ctx context.Context, seq ast.Seq, source *data.Source, head *lakeparse.Commitish) (dag.Seq, error) {
-	return newAnalyzer(ctx, source, head).semSeq(seq)
+	a := newAnalyzer(ctx, source, head)
+	s, err := a.semSeq(seq)
+	if err != nil {
+		return nil, err
+	}
+	op, err := a.buildFrom(s[0])
+	if err != nil {
+		return nil, err
+	}
+	if op != nil {
+		s.Prepend(op)
+	}
+	return s, nil
 }
 
 type analyzer struct {
@@ -47,6 +60,37 @@ func (a *analyzer) enterScope() {
 
 func (a *analyzer) exitScope() {
 	a.scope = a.scope.parent
+}
+
+func (a *analyzer) buildFrom(op dag.Op) (dag.Op, error) {
+	switch op := op.(type) {
+	case *dag.FileScan, *dag.HTTPScan, *dag.PoolScan, *dag.LakeMetaScan, *dag.PoolMetaScan, *dag.CommitMetaScan, *dag.DeleteScan:
+		return nil, nil
+	case *dag.Fork:
+		return a.buildFrom(op.Paths[0][0])
+	case *dag.Scope:
+		return a.buildFrom(op.Body[0])
+	case *dag.UserOpCall:
+		return a.buildFrom(op.Body[0])
+	}
+	// No from so add a source.
+	if a.head == nil {
+		return &kernel.Reader{}, nil
+	}
+	pool := &ast.Pool{
+		Kind: "Pool",
+		Spec: ast.PoolSpec{
+			Pool: &ast.String{
+				Kind: "String",
+				Text: "HEAD",
+			},
+		},
+	}
+	ops, err := a.semPool(pool)
+	if err != nil {
+		return nil, err
+	}
+	return ops[0], nil
 }
 
 type opDecl struct {

--- a/runtime/op/ztests/user-from.yaml
+++ b/runtime/op/ztests/user-from.yaml
@@ -1,0 +1,26 @@
+script: |
+  zq -z -I test.zed
+
+inputs:
+  - name: test.zed
+    data: |
+      op fileA(): (
+        file A.zson | sort a 
+      )
+      fileA()
+  - name: A.zson
+    data: |
+      {a:5}
+      {a:1}
+      {a:7}
+      {a:3}
+      {a:8}
+  
+outputs:
+  - name: stdout
+    data: |
+      {a:1}
+      {a:3}
+      {a:5}
+      {a:7}
+      {a:8}

